feat: new extension API `WithExecCommand`

DeagleGross opened the pull request:
## Description

PR is an implementation of `WithExecCommand` extension (see `In Aspire.Hosting` in #10301), which allows to register a command to execute against a container resource (actually inside the container - for example `ls` (list files) in the `nginx` container as is shown in tests). `WithExecCommand` is similar to `WithCommand` because it annotates a container resource with the new `ResourceContainerExecCommandAnnotation`. It is very similar to existing `ResourceCommandAnnotation`.

There has to be a way to actually launch a command and it will be implemented in the follow-up PR for #10301, part `In the dashboard`), and new `ContainerExecService` exposes an API to do it. `WithCommand` has different semantics: since the "command" is basically a c# callback, we can know if execution of the "command" has completed successfully, and we can return it. In case of containers there is no reliable way to know how the command execution went, and the most important here is to fetch logs, so I've built API which returns an `IAsyncEnumerable<ContainerExecCommandOutput>` - basically a stream of logs.

On the opposite of the `WithCommand` here it is necessary to run the the `ContainerExec` inside of DCP to actually execute the command inside the container. _See implementation details below._

### Public API changes

```diff
namespace Aspire.Hosting.ApplicationModel;

+ public class ContainerExecCommandOutput
+ {
+   public required string Text { get; init; }
+   public required bool IsErrorMessage { get; init; }
+   public int? LineNumber { get; init; }
+ }

+ public class ResourceContainerExecCommandAnnotation : ResourceCommandAnnotationBase
+ {
+   public string Command { get; }
+   public string? WorkingDirectory { get; }
+ }
```

```diff
namespace Aspire.Hosting.Exec;

+ public class ContainerExecService
+ {
+   public async IAsyncEnumerable<ContainerExecCommandOutput> ExecuteCommandAsync(string resourceId, string commandName, [EnumeratorCancellation] CancellationToken cancellationToken = default);

+   public IAsyncEnumerable<ContainerExecCommandOutput> ExecuteCommandAsync(ContainerResource resource, string commandName, CancellationToken cancellationToken = default);
+ }
```

### Implementation

I decided to allow `IDcpExecutor` to do 2 operations:
```diff
interface IDcpExecutor
{
+    Task<AppResource> RunEphemeralResourceAsync(IResource ephemeralResource, CancellationToken cancellationToken);
+    void DeleteEphemeralResource(AppResource? ephemeralResource);
}
```

The idea is that we dont know what custom commands will be executed against the container. In #10301 we will also support running custom commands from the dashboard via the interaction with the user. So we not always have an opportunity to register the `ContainerExecutableResource` before DCP starts, so I've extended the possibilities of `DcpExecutor` to run a new resource and to delete the handlers to it (only for monitoring purposes). I know `DcpExecutor` was an immutable entity before, so this can be a concerning change.

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [x] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes, will add later once full feature is complete

DeagleGross committed:
init

DeagleGross committed:
impl?

DeagleGross committed:
cleanup resources as well

DeagleGross committed:
merge main

DeagleGross committed:
cleanup ContainerExecsMap as well

DeagleGross commented on src/Aspire.Hosting/ApplicationModel/ResourceContainerExecCommandAnnotation.cs:
```diff
@@ -0,0 +1,64 @@
+public sealed class ResourceContainerExecCommandAnnotation : ResourceCommandAnnotationBase
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ResourceCommandAnnotation"/> class.
```
fixed

DeagleGross committed:
address PR comments x1

mitchdenny commented on src/Aspire.Cli/Backchannel/AppHostBackchannel.cs:
```diff
@@ -196,13 +196,13 @@ await rpc.InvokeWithCancellationAsync(
     }
 
-    public async IAsyncEnumerable<CommandOutput> ExecAsync([EnumeratorCancellation] CancellationToken cancellationToken)
+    public async IAsyncEnumerable<BackchannelCommandOutput> ExecAsync([EnumeratorCancellation] CancellationToken cancellationToken)
```
This will break compat.

DamianEdwards replied:
With what? This is an internal interface right? With older CLI instances?

DeagleGross replied:
This _can_ break compat, but for sure when updating the CLI nothing will be broken. I can revert naming change here - the idea was to distinguish an internal backchannel communication interface and the exposed type for getting logs via `ContainerExecService`

DeagleGross replied:
i reverted the naming change.

mitchdenny commented on src/Aspire.Hosting/ApplicationModel/ResourceCommandAnnotation.cs:
```diff
@@ -9,7 +9,7 @@ namespace Aspire.Hosting.ApplicationModel;
 /// </summary>
 [DebuggerDisplay("Type = {GetType().Name,nq}, Name = {Name}")]
-public sealed class ResourceCommandAnnotation : IResourceAnnotation
+public sealed class ResourceCommandAnnotation : ResourceCommandAnnotationBase
```
We shouldn't need to do this. The annotation provides a callback that should be able to capture everything it needs in the closure.

DeagleGross replied:
We have a different API here - we dont want any callback be provided by the user, so it should be a different annotation. I can avoid using a base type here, but callback is definitely not suitable here

mitchdenny commented on src/Aspire.Hosting/Exec/ContainerExecService.cs:
```diff
@@ -0,0 +1,153 @@
+/// <summary>
+/// A service to execute container exec commands.
+/// </summary>
+public class ContainerExecService
```
This should be an interface with an internal implementation I think.

mitchdenny commented on src/Aspire.Hosting/ResourceBuilderExtensions.cs:
```diff
@@ -1544,6 +1544,40 @@ public static IResourceBuilder<T> WithCommand<T>(
+    /// <param name="command">The command string to be executed.</param>
+    /// <param name="commandOptions">Optional settings for the command, such as description and icon.</param>
+    /// <returns>The resource builder, allowing for method chaining.</returns>
+    public static IResourceBuilder<T> WithExecCommand<T>(
```
I'm not sure about ExecCommand as the name.

DeagleGross replied:
as per requirement in the issue (see https://github.com/dotnet/aspire/issues/10301), it should be with the name.
It allows to add multiple commands per name to the same container resource. for example
```
builder
  .AddResource(containerResource)
  .WithExecCommand("listfiles", ...)
  .WithExecCommand("deletefiles", ...)
```

Why are you not sure about such API? It is basically a copy-cat from `WithCommand`

mitchdenny commented on tests/Aspire.Hosting.Tests/Backchannel/Exec/ExecTestsBase.cs:
```diff
@@ -43,7 +43,37 @@ internal async Task<List<CommandOutput>> ExecWithLogCollectionAsync(
-    internal static void AssertLogsContain(List<CommandOutput> logs, params string[] expectedLogMessages)
+    protected async Task<List<ContainerExecCommandOutput>> ProcessAndCollectLogs(IAsyncEnumerable<ContainerExecCommandOutput> containerExecLogs)
+    {
+        var logs = new List<ContainerExecCommandOutput>();
```
How will this work with long running operations with lots of log entries.

DeagleGross replied:
this exists only for tests and to easily check the collected logs as collection

DeagleGross committed:
revert backchannel type rename

karolz-ms commented:
(from the PR description)

> In case of containers there is no reliable way to know how the command execution went, and the most important here is to fetch logs, so I've built API which returns an IAsyncEnumerable<ContainerExecCommandOutput> - basically a stream of logs.

Not sure what you mean. The status for `ContainerExec` object has a `State` property and an `ExitCode` property https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/Model/ContainerExec.cs#L56 so you can know extactly what the outcome of an container executable was, once it has completed.

karolz-ms reviewed (changes requested):
My main concern is the interface for executing a container command, it needs to be re-designed IMO.

mitchdenny commented on src/Aspire.Hosting/DistributedApplicationBuilder.cs:
```diff
@@ -245,6 +245,7 @@ public DistributedApplicationBuilder(DistributedApplicationOptions options)
         _innerBuilder.Services.AddSingleton<ResourceNotificationService>();
         _innerBuilder.Services.AddSingleton<ResourceLoggerService>();
         _innerBuilder.Services.AddSingleton<ResourceCommandService>(s => new ResourceCommandService(s.GetRequiredService<ResourceNotificationService>(), s.GetRequiredService<ResourceLoggerService>(), s));
+        _innerBuilder.Services.AddSingleton<ContainerExecService>(s => new ContainerExecService(s.GetRequiredService<ResourceNotificationService>(), s.GetRequiredService<ResourceLoggerService>(), s.GetRequiredService<IDcpExecutor>(), s.GetRequiredService<DcpNameGenerator>()));
```
Why not just register the singleton - you shouldn't need the callback to construct the type here.

DeagleGross replied:
changed registration

mitchdenny commented:
OK, so I've reviewed this code. I definitely do not want us to introduce `ResourceCommandAnnotationBase`. I see this as unnecessary. I think the reason it feels necessary is that we've got an inverted relationship between `ResourceContainerExecCommandAnnotation` and `ContainerExecService`.

What we should be doing is in the `WithExecCommand(...)` method we should be invoking `WithCommand(...)` and into the callback we should be making a call to the DI container to resolve the `ContainerExecService` (or its interface) and then using that to invoke the command in the selected container resources. The parameters that are used to invoke the API on `ContainerExecService` would be passed in to the `WithExecCommand(...)` and captured in the lambda that is passed into the `WithCommand(...)` extension method.

We should look to `WithHttpCommand(...)` for inspiration here. the `WithExecCommand(...)` should probably have the essential arguments that we need for invoking a command inside the container - plus a derived `ExecCommandOptions` type.

One thing that we should probably think carefully about here is whether it actually makes sense to define container exec commands in the app model as a new kind of resource. Presently we have `ContainerResources` - perhaps we should have a `ContainerExecutableResource` which derives from `ExecutableResource`.

This way we might be able to take advantage of all of the plumbing we already have in Aspire around handling log streaming etc.

I believe that `aspire exec` will need to be converted with `aspire run` and that when doing `aspire exec` we will need to show the dashboard. We should probably resolve this before going too much further.

DeagleGross commented on src/Aspire.Hosting/Dcp/DcpExecutor.cs:
```diff
@@ -1877,6 +1883,38 @@ async Task EnsureResourceDeletedAsync<T>(string resourceName) where T : CustomRe
+            default: throw new InvalidOperationException($"Resource '{ephemeralResource.Name}' is not supported to run dynamically.");
+        }
+    }
+    public void DeleteEphemeralResource(AppResource? ephemeralResource)
```
made it non-nullable

DeagleGross commented on src/Aspire.Hosting/Dcp/IDcpExecutor.cs:
```diff
@@ -10,4 +12,7 @@ internal interface IDcpExecutor
     Task StopResourceAsync(IResourceReference resourceReference, CancellationToken cancellationToken);
+
+    Task<AppResource> RunEphemeralResourceAsync(IResource ephemeralResource, CancellationToken cancellationToken);
+    void DeleteEphemeralResource(AppResource? ephemeralResource);
```
changed to `Task DeleteEphemeralResourceAsync(AppResource ephemeralResource)`

DeagleGross commented on src/Aspire.Hosting/Exec/ContainerExecService.cs:
```diff
@@ -0,0 +1,153 @@
+        {
+            await _resourceNotificationService.WaitForResourceAsync(containerExecResource.Name, targetStates: KnownResourceStates.TerminalStates, cancellationToken).ConfigureAwait(false);
+
+            // hack: https://github.com/dotnet/aspire/issues/10245
```
i will remove it once code is refactored, thanks

DeagleGross committed:
address PR comments x1

DeagleGross committed:
stabilize and refactor

DeagleGross commented:
@mitchdenny thanks for the explanation about similarities with `WithCommand`, but I disagree - the API is simply different. In case of container exec underlying execution happens in a separate DCP resource, and we should have all the infra working for us providing a way to fetch logs for example. Trying to pack everything into `WithCommand` with its hardly binded `Func<ExecuteCommandContext, Task<ExecuteCommandResult>>` contract is not extensible enough. Please see the updated code and the idea by Karol here: https://github.com/dotnet/aspire/pull/10779#discussion_r2246103452.

However, I've refactored the code and removed the base annotation class etc, so its cleaner now.

The cli changes for `aspire exec` to be based on top of `aspire run` is unrelated to this flow, and should come in a separate PR. I will also definitely fix what you have described in other PR (https://github.com/dotnet/aspire/pull/10380#discussion_r2246837446_
 https://github.com/dotnet/aspire/pull/10380#discussion_r2246840159_); but will do in a separate PR for simplicity, if you are fine with that - this PR should not make it harder, because it provides a very simple command definition and args are not touched here

karolz-ms commented on src/Aspire.Hosting/Dcp/DcpExecutor.cs:
```diff
@@ -1877,6 +1883,38 @@ async Task EnsureResourceDeletedAsync<T>(string resourceName) where T : CustomRe
+
+        if (ephemeralResource.ModelResource is ContainerExecutableResource)
+        {
+            _resourceState.Remove(ephemeralResource);
```
Where is that comment? I do not see any in the latest version of `DeleteEphemeralResourceAsync()` method?

karolz-ms commented on src/Aspire.Hosting/Exec/ContainerExecService.cs:
```diff
@@ -0,0 +1,146 @@
+
+        Func<CancellationToken, Task<ExecuteCommandResult>> commandResultTask = async (CancellationToken cancellationToken) =>
+        {
+            await _dcpExecutor.RunEphemeralResourceAsync(containerExecResource, cancellationToken).ConfigureAwait(false);
```
It seems awkward that the call to `ContainerExecService.ExecuteCommand()` does not, in fact, executes the command. 

Consider making this method an async method and taking the call to `RunEphemenralResourceAsync()` out of the result-producing task. This will guarantee that when `ExecuteCommandCore()` returns without exception, the command is started. The returned run object would then expose `GetResultAsync()` method for waiting on the result. 

This change also guarantees that calling `GetOutputStream()` on the returned run object always makes sense. Without it, it is possible to ask for output stream without actually starting the command, which will result in waiting forever (until the cancellation token expires)

DamianEdwards commented:
>Personally, I am not very sure if modeling "execution of programs within containers" (avoiding the overloaded term "command" here) is a good candidate for being a distinct resource type in Aspire model. I am wondering if something more lightweight (e.g. just a method on the container resource) would ultimately be easier to use and more appropriate.

This is exactly my thinking too. I do not see this as a resource at all, but rather exposing the ability to execute/invoke something in the context of a resource.

mitchdenny commented:
My concern is that if we don't model this as a resource from top to bottom then we are going to end up with unnecessary duplications for something that conceptually is very well aligned with resources in the Aspire app model.

I mean we define models in Ollama as resources - why not a `docker exec` invocation? We've found that even for resources that aren't executables that there is a need to be able to inspect the output that they produce - it just makes sense that we would use the existing dashboard infrastructure for this doesn't it?
